File: generate/image_to_pc.py
from PIL import Image
import numpy as np
import sys

DIMENSIONS = 30
THRESHOLD = 128


def main():
    if len(sys.argv) != 2:
        raise ValueError("Usage: python image_to_pc.py <image>")

    # Open an image
    img = Image.open(sys.argv[1])

    # Ensure height equals width
    if img.width != img.height:
        raise ValueError("Image must be square")

    # Crop image so size if multiple of DIMENSIONS
    new_size = img.width - (img.width % DIMENSIONS)
    img = img.crop((0, 0, new_size, new_size))

    # Build blocks
    blocks = createBlocks(img)
    print(blocks)
    writeBlocksToFile(blocks)


def createBlocks(img):
    pixels = img.load()
    block_size = img.width // DIMENSIONS

    # Init numpy array representing blocks
    blocks = np.zeros((DIMENSIONS, DIMENSIONS))

    # Iterate over blocks
    for i in range(0, img.width, block_size):
        for j in range(0, img.height, block_size):
            # Iterate over pixels in block
            avg = 0
            for x in range(i, i + block_size):
                for y in range(j, j + block_size):
                    a = pixels[x, y]

                    if a > THRESHOLD:
                        avg += 1

            # Color pixel depending on average of pixels
            avg /= block_size**2
            avg = round(avg)
            color = 255 if avg == 1 else 0
            blocks[i // block_size, j // block_size] = color == 255

            # Color pixels depending on average
            # for x in range(i, i + block_size):
            #     for y in range(j, j + block_size):
            #         pixels[x, y] = color
            # img.show()
    return blocks


def writeBlocksToFile(blocks):
    with open("out.pc", "w") as file:
        colLine = ""
        rowLine = ""

        for i in range(DIMENSIONS):
            row = blocks[i, :]
            col = blocks[:, i]

            lineNumber = 0
            colNumber = 0
            for j in range(DIMENSIONS):
                if row[j]:
                    lineNumber += 1
                elif lineNumber != 0:
                    colLine += f"{lineNumber} "
                    lineNumber = 0

                if col[j]:
                    colNumber += 1
                elif colNumber != 0:
                    rowLine += f"{colNumber} "
                    colNumber = 0

            if lineNumber != 0:
                colLine += f"{lineNumber} "
            if colNumber != 0:
                rowLine += f"{colNumber} "

            if i != DIMENSIONS - 1:
                colLine += "| "
                rowLine += "| "

        file.write(colLine + "\n")
        file.write(rowLine + "\n")


if __name__ == "__main__":
    main()

File: src/bin/image_to_pc.rs
use clap::Parser;
use image::DynamicImage;
use std::path::PathBuf;
use std::vec;
use std::fs::File;
use std::io::Write;
use image::GenericImageView;

const THRESHOLD: u8 = 128;

#[derive(Parser)]
struct Args {
    image: PathBuf,
    dimension: u32
}


fn main() -> anyhow::Result<()> {
    let args: Args = Args::parse();

    let mut img: DynamicImage = image::open(&args.image).expect("Failed to open image");

    if img.width() != img.height() {
        panic!("The provided image should be a square image!");
    }
    
    // Crop image if not multiple of dimension
    let new_size = img.width() - (img.width() % args.dimension);
    img = img.crop(0, 0, new_size, new_size);

    let blocks: Vec<Vec<bool>> = create_blocks(&img, &args.dimension);

    // Open and write to files
    let image_file: String = args.image.file_stem().expect("Path cannot be empty").to_str().unwrap().to_owned();
    write_blocks_to_file(&blocks, &image_file, &args.dimension);    
    write_solutions(&blocks, &image_file, &args.dimension);
    Ok(())
}


fn create_blocks(img: &DynamicImage, dimension: &u32) -> Vec<Vec<bool>> {

    let block_size : u32 = img.width() / *dimension;
    let mut blocks : Vec<Vec<bool>> = vec![vec![false; *dimension as usize]; *dimension as usize];

    for i in (0 .. img.width()).step_by(block_size as usize) {
        for j in (0..img.height()).step_by(block_size as usize) {
            let mut avg: f64 = 0.0;

            for x in i .. i+block_size {
                for y in j .. j+block_size{
                    let pixel = img.get_pixel(x, y);
                    let pixel_average = (pixel.0[0] as f64 + pixel.0[1] as f64 + pixel.0[2] as f64) / 3.0;
                    if pixel_average < THRESHOLD as f64 {
                        avg += 1.0;
                    }
                }
            }
            

            avg = avg / (block_size.pow(2) as f64);
            avg = avg.round();
            
            let color : u8;
            if avg == 1.0 {
                color = 255
            }
            else {
                color = 0
            }

            blocks[(i / block_size) as usize][(j / block_size) as usize] = color == 255;

            for _ in i .. i+block_size {
                for _ in j .. j+block_size{
                    img.get_pixel(i, j).0 = [color, color, color, 255];
                }
            }
        }
    }
    println!("Blocks: {:?}", blocks);
    return blocks;
}

fn write_blocks_to_file(blocks: &Vec<Vec<bool>>, filename: &String, dimension: &u32) {

    let full_filename = format!("{}.pc", filename);
    let mut file = File::create(full_filename).expect("Failed to create file");

    let mut column_line = String::new();
    let mut row_line = String::new();

    for i in 0..*dimension {
        let row = blocks[i as usize].iter().map(|x| x).collect::<Vec<&bool>>();
        let column = blocks.iter().map(|x| x[i as usize]).collect::<Vec<bool>>();

        let mut row_number = 0;
        let mut column_number = 0;

        for j in 0..*dimension {
            
            // Append row line
            if *row[j as usize] {
                row_number += 1;
            } else if row_number != 0 {
                row_line.push_str(&format!("{} ", row_number));
                row_number = 0;
            }

            // Append column line
            if column[j as usize] {
                column_number += 1;
            } else if column_number != 0 {
                column_line.push_str(&format!("{} ", column_number));
                column_number = 0;
            }
        }
        if column_number != 0 {
            column_line.push_str(&format!("{} ", column_number));
        } 
        if row_number != 0 {
            row_line.push_str(&format!("{} ", row_number));
        }

        if i != dimension - 1 {
            column_line.push_str("| ");
            row_line.push_str("| ");
        }
    }
    let full_line = format!("{}\n{}\n", row_line, column_line);
    file.write_all(full_line.as_bytes()).expect("Failed to write to file.");

}

fn write_solutions(blocks: &Vec<Vec<bool>>, filename: &String, dimension: &u32) {
    let filepath = format!("{}.pcs", filename);
    let mut file = File::create(filepath).expect("Failed to create file");
    for i in 0..*dimension{
        for j in 0..*dimension{
            let ch = if blocks[j as usize][i as usize] { '#' } else { '.' };
            write!(file, "{}",ch).expect("Failed to write to file");
        }
        writeln!(file).expect("Failed to write newline")
    }    
}